### PR TITLE
feat: Create inital (de-)serialization types

### PR DESCRIPTION
**Pull request by marikaner**
Related to https://github.com/SAP/cloud-sdk-backlog/issues/399

This PR provides the types necessary to build (de-)serializers.
This uses a set of default deserializers, but does not yet allow to use custom (de-)serializers.

### CODE REVIEW

File: packages/odata-common/src/de-serializers/de-serializers.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * Represents a set of functions for serialization, deserialization and optionally serialization for URIs.
+ * The return type of the `deserialize` function is the type of the input to the `serialize` and `serializeToUri` functions.
+ */
+export interface DeSerializer<DeserializedT> {
+  /**
+   * Function to deserialize a value.
+   * @param val - Serialized value to deserialize.
+   * @returns Deserialized value.
+   */
+  deserialize: (val: any) => DeserializedT;
```
Comment (florian-richter) [inline packages/odata-common/src/de-serializers/de-serializers.ts:48]:
```suggestion
  /**
   * Function to deserialize a value which is _usually_ a string.
   * @param value - Serialized value to deserialize.
   * @returns Deserialized value.
   */
  deserialize: (value: any) => DeserializedT;
```

File: packages/odata-common/src/de-serializers/de-serializers.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  /**
+   * Function to serialize a value.
+   * @param val - Deserialized value to serialize.
+   * @returns Serialized value.
+   */
+  serialize: (val: DeserializedT) => any;
```
Comment (florian-richter) [inline packages/odata-common/src/de-serializers/de-serializers.ts:55]: Should `any` be `string`? Should we explain why or why not?

```suggestion
  /**
   * Function to serialize a value.
   * @param value - Deserialized value to serialize.
   * @returns Serialized value.
   */
  serialize: (value: DeserializedT) => any;
```

File: packages/odata-common/src/de-serializers/de-serializers.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+   * Function to serialize a value for use in a URI.
+   * This function is optional. If not specified, the `serialize` function will be used.
+   * @param val - Deserialized value to serialize.
+   * @param serialize - The `serialize` function to serialize the payload.
+   * @returns Serialized value for URIs.
+   */
+  serializeToUri?: (
+    val: DeserializedT,
+    serialize: DeSerializer<DeserializedT>['serialize']
+  ) => string;
```
Comment (florian-richter) [inline packages/odata-common/src/de-serializers/de-serializers.ts:67]:
```suggestion
  /**
   * Function to serialize a value for use in a URI.
   * This function is optional. If not specified, the `serialize` function will be used.
   * @param value - Deserialized value to serialize.
   * @param serialize - A reference to the `serialize` function which is usually used to serialize the payload.
   * @returns Serialized value for URIs.
   */
  serializeToUri?: (
    value: DeserializedT,
    serialize: DeSerializer<DeserializedT>['serialize']
  ) => string;
```

File: packages/odata-common/src/de-serializers/de-serializers.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+/**
+ * Infers the deserialized type for an EDM type from the given `DeSerializers` type.
+ * @typeparam DeSerializersT - Type of the `DeSerializers`.
+ * @typeparam EdmT - EDM type to get the deserialized type for.
```
Comment (florian-richter) [inline packages/odata-common/src/de-serializers/de-serializers.ts:73]:
```suggestion
 * @typeparam EdmT - Return type of the deserialize function for the given EDM type.
```

File: packages/odata-common/src/de-serializers/de-serializers.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+> = EdmT extends keyof DeSerializersT
+  ? DeSerializersT[EdmT] extends DeSerializer<infer DeserializedT>
+    ? DeserializedT
+    : never
+  : never;
+
+export const identityDeSerializerRaw = {
+  deserialize: identity,
+  serialize: identity
+};
```
Comment (florian-richter) [inline packages/odata-common/src/de-serializers/de-serializers.ts:87]: Remove?
```suggestion
```

File: packages/odata-common/src/de-serializers/default-de-serializers.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    serialize: identity,
+    serializeToUri: (value, serialize) =>
+      convertToUriForEdmString(serialize(value))
+  },
+  'Edm.Any': {
+    deserialize: identity,
+    serialize: identity,
+    serializeToUri: (value, serialize) => String(serialize(value))
+  },
+  /* DeSerializers without common URI serializer defaults. */
```
Comment (florian-richter) [inline packages/odata-common/src/de-serializers/default-de-serializers.ts:108]:
```suggestion
  /* DeSerializers with v2 and v4 specific URI serializer defaults. */
```

File: packages/odata-common/src/de-serializers/default-de-serializers.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    return value;
+  };
+}
+
+/**
+ * Wraps the given default deserialization function with a check for null values.
+ * @param deserialize - Deserialization function to wrap.
+ * @returns The wrapped deserialization function.
+ */
+export function wrapDeserialization<T>(
```
Comment (florian-richter) [inline packages/odata-common/src/de-serializers/default-de-serializers.ts:143]: Dont forget:
```
export function wrapDefaultDeserialization<T>(
```

File: packages/odata-v2/src/de-serializers/payload-value-converter.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
+/* eslint-disable valid-jsdoc */
```
Comment (florian-richter) [inline packages/odata-v2/src/de-serializers/payload-value-converter.ts:1]: Comment -> To be deleted once we get rid of emdToTs and tsToEdm

File: packages/odata-v4/src/de-serializers/default-de-serializers.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    serialize: serializeToTime,
+    serializeToUri: (value, serialize) => serialize(value)
+  },
+  'Edm.Enum': {
+    deserialize: identity,
+    serialize: identity,
+    serializeToUri: (value, serialize) =>
+      convertToUriForEdmString(serialize(value))
+  },
+  /* DeSerializers without common URI serializer defaults. */
```
Comment (florian-richter) [inline packages/odata-v4/src/de-serializers/default-de-serializers.ts:79]:
```suggestion
  /* DeSerializers with specific v2 and v4 URI serializer defaults. */
```

File: packages/odata-v4/src/de-serializers/payload-value-converter.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
+/* eslint-disable valid-jsdoc */
```
Comment (florian-richter) [inline packages/odata-v4/src/de-serializers/payload-value-converter.ts:1]: Comment "Delete once functions are removed" -> deprecate edmToTs? tsToEdm?

File: packages/odata-v2/src/de-serializers/default-de-serializers.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    deserialize: deserializeToMoment,
+    serialize: serializeFromMoment,
+    serializeToUri: value => `datetimeoffset'${value.toISOString()}'`
+  },
+  'Edm.Time': {
+    deserialize: deserializeToTime,
+    serialize: serializeFromTime,
+    serializeToUri: (value, serialize) => `time'${serialize(value)}'`
+  },
+  /* DeSerializers without common URI serializer defaults. */
```
Comment (marikaner) [inline packages/odata-v2/src/de-serializers/default-de-serializers.ts:60]:
```suggestion
/* DeSerializers with v2 specific URI serializer defaults. */
```

File: packages/odata-v4/src/de-serializers/payload-value-converter.ts (hunk 10 above)
Comment (marikaner) [reply packages/odata-v4/src/de-serializers/payload-value-converter.ts:1]:
```suggestion
/* eslint-disable valid-jsdoc */
/* Delete this file once the edmToTs and tsToEdm are obsolete. */
```

File: packages/connectivity/src/scp-cf/identity-service.ts [hunk 12]
```diff
@@ -1,6 +1,6 @@
 import * as xssec from '@sap/xssec';
 import { first } from '@sap-cloud-sdk/util';
-import { DestinationFetchOptions, DestinationOptions } from '..';
+import { DestinationFetchOptions, DestinationOptions } from './destination';
```
Comment (marikaner) [inline packages/connectivity/src/scp-cf/identity-service.ts:3]:
```suggestion
import { DestinationFetchOptions, DestinationOptions } from '..';
```

File: packages/http-client/src/csrf-token-header.ts [hunk 13]
```diff
@@ -12,7 +12,7 @@ import {
   DestinationFetchOptions
 } from '@sap-cloud-sdk/connectivity';
 import { HttpRequestConfig } from './http-client-types';
-import { executeHttpRequest } from '.';
+import { executeHttpRequest } from './http-client';
```
Comment (marikaner) [inline packages/http-client/src/csrf-token-header.ts:15]:
```suggestion
import { executeHttpRequest } from '.';
```

File: packages/odata-common/src/entity-base.ts [hunk 14]
```diff
@@ -3,10 +3,8 @@
 import { camelCase, equal, isNullish } from '@sap-cloud-sdk/util';
 import { EntityBuilder } from './entity-builder';
 import { isNavigationProperty, nonEnumerable } from './properties-util';
-import type { Field } from './selectable/field';
-import type { Link } from './selectable/link';
-import type { RequestBuilder } from './request-builder/request-builder';
-import type { CustomField } from './selectable/custom-field';
+import type { Field, Link, CustomField } from './selectable';
+import type { RequestBuilder } from './request-builder';
```
Comment (marikaner) [inline packages/odata-common/src/entity-base.ts:7]:
```suggestion
import type { Field } from './selectable/field';
import type { Link } from './selectable/link';
import type { RequestBuilder } from './request-builder/request-builder';
import type { CustomField } from './selectable/custom-field';
```

File: packages/odata-common/src/entity-deserializer.ts [hunk 15]
```diff
@@ -17,11 +17,13 @@ import {
...
 import { Field } from './selectable/field';
 import { Link } from './selectable/link';
-import { EdmTypeField } from './selectable/edm-type-field';
+import {
+  EdmTypeField,
+  CollectionField,
+  ComplexTypeField,
+  OneToOneLink
+} from './selectable';
 import { EnumField } from './selectable/enum-field';
```
Comment (marikaner) [inline packages/odata-common/src/entity-deserializer.ts:26]:
```suggestion
import { EdmTypeField } from './selectable/edm-type-field';
import { EnumField } from './selectable/enum-field';
import { CollectionField } from './selectable/collection-field';
import { ComplexTypeField } from './selectable/complex-type-field';
import { OneToOneLink } from './selectable/one-to-one-link';
```

File: packages/odata-common/src/uri-conversion/get-filter.ts [hunk 16]
```diff
@@ -4,22 +4,22 @@ import { Constructable, EntityBase } from '../entity-base';
...
+  Filterable,
+  FilterList,
+  isFilterList,
+  isBooleanFilterFunction,
+  FilterLink,
+  isFilterLink,
+  isUnaryFilter,
+  UnaryFilter,
+  Filter,
+  isFilter,
```
Comment (marikaner) [inline packages/odata-common/src/uri-conversion/get-filter.ts:17]:
```suggestion
  isFilterLambdaExpression
} from '../filter/filter-lambda-expression';
import { Filterable } from '../filter/filterable';
import { FilterList, isFilterList } from '../filter/filter-list';
import { isBooleanFilterFunction } from '../filter/boolean-filter-function';
import { FilterLink, isFilterLink } from '../filter/filter-link';
import { isUnaryFilter, UnaryFilter } from '../filter/unary-filter';
import { OneToManyLink } from '../selectable/one-to-many-link';
import { ComplexTypeField } from '../selectable/complex-type-field';
import {
```

File: packages/odata-common/src/uri-conversion/get-filter.ts [hunk 17]
```diff
@@ -4,22 +4,22 @@ import { Constructable, EntityBase } from '../entity-base';
...
   FilterFunction,
   FilterFunctionParameterType
-} from '../filter/filter-function-base';
-import { FieldType } from '../selectable/field';
-import { Filter, isFilter } from '../filter/filter';
-import { UriConverter } from './uri-value-converter';
+} from '../filter';
+import { OneToManyLink, ComplexTypeField, FieldType } from '../selectable';
```
Comment (marikaner) [inline packages/odata-common/src/uri-conversion/get-filter.ts:21]:
```suggestion
} from '../filter/filter-function-base';
import { FieldType } from '../selectable/field';
import { Filter, isFilter } from '../filter/filter';
```

File: packages/odata-common/src/uri-conversion/odata-uri.ts [hunk 18]
```diff
@@ -1,13 +1,12 @@
...
 import { EntityBase, Constructable } from '../entity-base';
 import { EdmTypeShared } from '../edm-types';
-import { Filterable } from '../filter/filterable';
-import { Selectable } from '../selectable/selectable';
-import { Orderable } from '../order/orderable';
-import { FieldType } from '../selectable/field';
+import { Selectable, FieldType } from '../selectable';
+import { Orderable } from '../order';
+import { Filterable } from '../filter';
+import { UriConverter } from '../de-serializers';
```
Comment (marikaner) [inline packages/odata-common/src/uri-conversion/odata-uri.ts:7]:
```suggestion
import { Filterable } from '../filter/filterable';
import { Selectable } from '../selectable/selectable';
import { Orderable } from '../order/orderable';
import { FieldType } from '../selectable/field';
```

File: packages/odata-v2/src/entity-deserializer.ts [hunk 19]
```diff
@@ -1,12 +1,10 @@
...
-  EntityDeserializer,
-  EdmTypeShared
+  EntityDeserializer
 } from '@sap-cloud-sdk/odata-common/internal';
-import { EdmToPrimitive, edmToTs } from './payload-value-converter';
+import { edmToTs } from './de-serializers';
 import { extractODataEtag } from './extract-odata-etag';
-import { getLinkedCollectionResult } from './request-builder/response-data-accessor';
-import { EdmType } from './edm-types';
+import { getLinkedCollectionResult } from './request-builder';
```
Comment (marikaner) [inline packages/odata-v2/src/entity-deserializer.ts:7]:
```suggestion
import { getLinkedCollectionResult } from './request-builder/response-data-accessor';
```

File: packages/odata-v2/src/index.ts [hunk 20]
```diff
@@ -8,24 +6,25 @@ export {
...
-export { deserializeComplexType } from './entity-deserializer';
+  transformReturnValueForUndefined,
+  CreateRequestBuilder,
+  DeleteRequestBuilder,
+  GetAllRequestBuilder,
+  GetByKeyRequestBuilder,
+  ODataBatchRequestBuilder,
+  UpdateRequestBuilder,
+  FunctionImportRequestBuilder
+} from './request-builder';
```
Comment (marikaner) [inline packages/odata-v2/src/index.ts:24]:
```suggestion
transformReturnValueForUndefined
} from './request-builder/response-transformers';
```

File: packages/odata-v2/src/index.ts [hunk 21]
```diff
@@ -8,24 +6,25 @@ export {
...
-export { CreateRequestBuilder } from './request-builder/create-request-builder';
-export { DeleteRequestBuilder } from './request-builder/delete-request-builder';
-export { GetAllRequestBuilder } from './request-builder/get-all-request-builder';
-export { GetByKeyRequestBuilder } from './request-builder/get-by-key-request-builder';
-export { ODataBatchRequestBuilder } from './request-builder/batch-request-builder';
-export { UpdateRequestBuilder } from './request-builder/update-request-builder';
-export { FunctionImportRequestBuilder } from './request-builder/function-import-request-builder';
-export { CustomField } from './selectable/custom-field';
 export { Entity } from './entity';
+export { CustomField } from './selectable';
```
Comment (marikaner) [inline packages/odata-v2/src/index.ts:27]:
```suggestion
```

File: packages/odata-v4/src/index.ts [hunk 22]
```diff
@@ -16,28 +13,29 @@ export {
...
-export { deserializeComplexType } from './entity-deserializer';
+  transformReturnValueForUndefined,
+  CreateRequestBuilder,
+  DeleteRequestBuilder,
+  GetAllRequestBuilder,
+  GetByKeyRequestBuilder,
+  UpdateRequestBuilder,
+  ActionImportRequestBuilder,
+  FunctionImportRequestBuilder
+} from './request-builder';
```
Comment (marikaner) [inline packages/odata-v4/src/index.ts:31]:
```suggestion
  transformReturnValueForUndefined
} from './request-builder/response-transformers';

export { CreateRequestBuilder } from './request-builder/create-request-builder';
export { DeleteRequestBuilder } from './request-builder/delete-request-builder';
export { GetAllRequestBuilder } from './request-builder/get-all-request-builder';
```
export { GetByKeyRequestBuilder } from './request-builder/get-by-key-request-builder';
export { UpdateRequestBuilder } from './request-builder/update-request-builder';

File: packages/odata-v4/src/entity-deserializer.ts [hunk 23]
```diff
@@ -1,12 +1,10 @@
...
-  EdmTypeShared,
   entityDeserializer as entityDeserializerBase,
   EntityDeserializer
 } from '@sap-cloud-sdk/odata-common/internal';
-import { EdmToPrimitive, edmToTs } from './payload-value-converter';
+import { edmToTs } from './de-serializers';
 import { extractODataEtag } from './extract-odata-etag';
-import { getLinkedCollectionResult } from './request-builder/response-data-accessor';
-import { EdmType } from './edm-types';
+import { getLinkedCollectionResult } from './request-builder';
```
Comment (marikaner) [inline packages/odata-v4/src/entity-deserializer.ts:7]:
```suggestion
import { getLinkedCollectionResult } from './request-builder/response-data-accessor';
```

File: packages/odata-v4/src/batch-request-builder.ts [hunk 24]
```diff
@@ -9,7 +9,7 @@ import {
   BatchRequestBuilder,
   deserializeBatchResponse
 } from '@sap-cloud-sdk/odata-common/internal';
-import { responseDataAccessor } from './request-builder/response-data-accessor';
+import { responseDataAccessor } from './request-builder';
```
Comment (marikaner) [inline packages/odata-v4/src/batch-request-builder.ts:12]:
```suggestion
import { responseDataAccessor } from './request-builder/response-data-accessor';```

File: test-packages/type-tests/package.json [hunk 25]
```diff
@@ -13,6 +13,7 @@
   },
   "dependencies": {
     "@sap-cloud-sdk/http-client": "^1.50.0",
+    "@sap-cloud-sdk/odata-v2": "^1.50.0",
```
Comment (marikaner) [inline test-packages/type-tests/package.json:16]:
```suggestion
    "@sap-cloud-sdk/odata-v2": "^1.50.0",
    "@sap-cloud-sdk/odata-v4": "^1.50.0",
```

**Review by florian-richter — APPROVED** [review]
LGTM

**Pull request merged**